In [54]:
import pandas as pd

Load the data and combine it into one dataframe.

In [55]:
root_folder = "./Data"
jun24 = pd.read_csv(f'{root_folder}/202406-divvy-tripdata.csv', index_col=0)
jul24 = pd.read_csv(f'{root_folder}/202407-divvy-tripdata.csv', index_col=0)
aug24 = pd.read_csv(f'{root_folder}/202408-divvy-tripdata.csv', index_col=0)
sept24 = pd.read_csv(f'{root_folder}/202409-divvy-tripdata.csv', index_col=0)
oct24 = pd.read_csv(f'{root_folder}/202410-divvy-tripdata.csv', index_col=0)
nov24 = pd.read_csv(f'{root_folder}/202411-divvy-tripdata.csv', index_col=0)
dec24 = pd.read_csv(f'{root_folder}/202412-divvy-tripdata.csv', index_col=0)
jan25 = pd.read_csv(f'{root_folder}/202501-divvy-tripdata.csv', index_col=0)
feb25 = pd.read_csv(f'{root_folder}/202502-divvy-tripdata.csv', index_col=0)
mar25 = pd.read_csv(f'{root_folder}/202503-divvy-tripdata.csv', index_col=0)
apr25 = pd.read_csv(f'{root_folder}/202504-divvy-tripdata.csv', index_col=0)
may25 = pd.read_csv(f'{root_folder}/202505-divvy-tripdata.csv', index_col=0)

In [56]:
data = pd.concat([jun24, jul24, aug24, sept24, oct24, nov24, dec24, jan25, feb25, mar25, apr25, may25])
data = data.reset_index(level=0)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5628847 entries, 0 to 5628846
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 558.3+ MB


In [57]:
data.sample(10)

ride_id     rideable_type               started_at  \
5517312  9382D0E0FE045DD8     electric_bike  2025-05-29 15:56:29.109   
3057920  CAAB81894854B932     electric_bike  2024-10-20 13:43:45.471   
985496   1750EDCA33D1E04F     electric_bike  2024-07-22 15:40:32.313   
612908   0E209BA63EF4221D      classic_bike  2024-06-12 13:35:17.005   
2337509  D69915B7BD127143     electric_bike  2024-09-03 18:37:19.773   
3014912  3A3F545234BAC728     electric_bike  2024-09-09 15:53:40.381   
4573874  5B3BB196B2828C29     electric_bike  2025-03-21 11:58:34.616   
2578910  81410EC8297E187E     electric_bike  2024-09-03 06:38:32.902   
1421518  357910CAE3718E48      classic_bike  2024-07-27 12:26:16.285   
2898127  A8BFF09C0F38C6A8  electric_scooter  2024-09-26 21:42:12.967   

                        ended_at             start_station_name  \
5517312  2025-05-29 16:02:29.064  Ravenswood Ave & Lawrence Ave   
3057920  2024-10-20 13:50:43.342                            NaN   
985496   2024-07-22 15:56:58.951  Milwaukee Ave & Fullerton Ave   
612908   2024-06-12 13:52:22.961          Michigan Ave & 8th St   
2337509  2024-09-03 19:30:35.555            Theater on the Lake   
3014912  2024-09-09 16:11:54.931                            NaN   
4573874  2025-03-21 12:05:45.934                            NaN   
2578910  2024-09-03 06:47:36.470    Lincoln Ave & Fullerton Ave   
1421518  2024-07-27 12:57:44.007       Kingsbury St & Kinzie St   
2898127  2024-09-26 21:56:14.721    Lincoln Ave & Fullerton Ave   

        start_station_id              end_station_name end_station_id  \
5517312     TA1309000066     Western Ave & Berteau Ave            438   
3057920              NaN                           NaN            NaN   
985496               428                           NaN            NaN   
612908               623  Desplaines St & Jackson Blvd          15539   
2337509     TA1308000001      Fairbanks Ct & Grand Ave   TA1305000003   
3014912              NaN                           NaN            NaN   
4573874              NaN       Paulina Ave & North Ave   TA1305000037   
2578910     TA1309000058             Wells St & Elm St   KA1504000135   
1421518     KA1503000043           Theater on the Lake   TA1308000001   
2898127     TA1309000058    Larrabee St & Armitage Ave   TA1309000006   

         start_lat  start_lng    end_lat    end_lng member_casual  
5517312  41.968445 -87.674226  41.957740 -87.688273        member  
3057920  41.900000 -87.620000  41.880000 -87.620000        member  
985496   41.924691 -87.700550  41.890000 -87.650000        member  
612908   41.872773 -87.623981  41.878119 -87.643948        member  
2337509  41.926277 -87.630834  41.891847 -87.620580        casual  
3014912  41.870000 -87.650000  41.910000 -87.640000        member  
4573874  41.900000 -87.670000  41.909855 -87.669930        member  
2578910  41.924161 -87.646380  41.903222 -87.634324        casual  
1421518  41.889177 -87.638506  41.926277 -87.630834        casual  
2898127  41.924161 -87.646380  41.918084 -87.643749        casual

Change the datatype of datetime columns

In [58]:
data['started_at'] = pd.to_datetime(data['started_at'])
data['ended_at'] = pd.to_datetime(data['ended_at'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5628847 entries, 0 to 5628846
Data columns (total 13 columns):
 #   Column              Dtype         
---  ------              -----         
 0   ride_id             object        
 1   rideable_type       object        
 2   started_at          datetime64[ns]
 3   ended_at            datetime64[ns]
 4   start_station_name  object        
 5   start_station_id    object        
 6   end_station_name    object        
 7   end_station_id      object        
 8   start_lat           float64       
 9   start_lng           float64       
 10  end_lat             float64       
 11  end_lng             float64       
 12  member_casual       object        
dtypes: datetime64[ns](2), float64(4), object(7)
memory usage: 558.3+ MB


The values of rideable_type should be unique to "electric_bike", "classic_bike" and "electric_scooter"

In [59]:
data.rideable_type.unique()

array(['electric_bike', 'classic_bike', 'electric_scooter'], dtype=object)

The value of member_casual should be "casual" or "member"

In [60]:
data.member_casual.unique()

array(['casual', 'member'], dtype=object)

Check if ride_id is unique

In [61]:
data.ride_id.nunique() == data.ride_id.count()

True

Check the occurrences of start station and end station

In [62]:
data.start_station_id.value_counts()

start_station_id
13022           60785
KA1503000043    41467
13300           40776
13042           37029
LF-005          37010
                ...  
607                 1
625                 1
668                 1
726                 1
606                 1
Name: count, Length: 1785, dtype: int64

In [63]:
data.end_station_id.value_counts()

end_station_id
13022           62041
KA1503000043    40908
LF-005          40744
13300           39451
13042           36935
                ...  
1244.0              1
707                 1
769                 1
895                 1
1083.0              1
Name: count, Length: 1787, dtype: int64

In [76]:
sid = data['start_station_id'].value_counts()
sid[sid== 1].count()

77

There are 77 stations that have appeared just once, it can mean that these stations are not used.

In [65]:
eid = data['end_station_id'].value_counts()
eid[eid== 1].count()

76

There are 76 end stations that have appeared just once, it can mean that these stations are not used.

In [66]:
data.isnull().sum()

ride_id                     0
rideable_type               0
started_at                  0
ended_at                    0
start_station_name    1081945
start_station_id      1081945
end_station_name      1110740
end_station_id        1110740
start_lat                   0
start_lng                   0
end_lat                  6166
end_lng                  6166
member_casual               0
dtype: int64

Dropping all the rows with null values

In [67]:
data_no_null = data.dropna()
data_no_null.shape

(3959660, 13)

Creating ride time column in seconds

In [68]:
df_ride = data_no_null.copy(deep=True)

df_ride['ride_time[s]'] = (df_ride['ended_at'] - df_ride['started_at']).dt.total_seconds()

df_ride.sample(10)

ride_id  rideable_type              started_at  \
1711206  A2ABF2C8499AC090   classic_bike 2024-08-03 17:22:46.288   
1856538  BE0700D9DBB54CAC  electric_bike 2024-08-02 07:50:42.057   
2730535  5825D37EF9EE9D1A   classic_bike 2024-09-12 09:56:20.206   
369839   13F978391EC0D844  electric_bike 2024-06-10 13:56:58.893   
5542980  E7413786B5DA1492  electric_bike 2025-05-16 12:00:31.919   
820333   1571F27EFC318614   classic_bike 2024-07-06 17:15:37.032   
5034074  1162726251CABD0D  electric_bike 2025-04-20 19:59:05.239   
3439725  1448420DABC86636  electric_bike 2024-10-18 09:35:12.050   
4820877  5E029A168E7EB5DD   classic_bike 2025-04-22 17:31:57.726   
3249873  7A6FEDEC6FBF057D   classic_bike 2024-10-12 13:55:01.795   

                       ended_at               start_station_name  \
1711206 2024-08-03 17:39:07.971            Wells St & Concord Ln   
1856538 2024-08-02 07:54:35.492   Southport Ave & Wrightwood Ave   
2730535 2024-09-12 10:06:33.355          Halsted St & Maxwell St   
369839  2024-06-10 14:03:55.807           Halsted St & Fulton St   
5542980 2025-05-16 12:04:00.534        Aberdeen St & Randolph St   
820333  2024-07-06 18:34:57.438  Cityfront Plaza Dr & Pioneer Ct   
5034074 2025-04-20 20:01:19.409       Kedzie Ave & Milwaukee Ave   
3439725 2024-10-18 09:40:10.474             Paulina St & 18th St   
4820877 2025-04-22 17:40:43.385           Kingsbury St & Erie St   
3249873 2024-10-12 14:18:51.123            Calumet Ave & 18th St   

        start_station_id               end_station_name end_station_id  \
1711206     TA1308000050          Dearborn St & Erie St          13045   
1856538     TA1307000113  Sheffield Ave & Fullerton Ave   TA1306000016   
2730535     TA1309000001       Racine Ave & Randolph St          13155   
369839             23003          Wells St & Hubbard St   TA1307000151   
5542980            18062     Aberdeen St & Jackson Blvd          13157   
820333             13427     Franklin St & Jackson Blvd   TA1305000025   
5034074            13085  St. Louis Ave & Fullerton Ave   KA1504000090   
3439725     TA1307000159           Halsted St & 18th St          13099   
4820877            13265            Canal St & Adams St          13011   
3249873            13102      Columbus Dr & Randolph St          13263   

         start_lat  start_lng    end_lat    end_lng member_casual  \
1711206  41.912133 -87.634656  41.893992 -87.629318        casual   
1856538  41.928773 -87.663913  41.925602 -87.653708        member   
2730535  41.864883 -87.647071  41.884211 -87.654254        member   
369839   41.886814 -87.648108  41.889906 -87.634266        casual   
5542980  41.884114 -87.654264  41.877726 -87.654787        casual   
820333   41.890573 -87.622072  41.877708 -87.635321        casual   
5034074  41.928999 -87.708030  41.924816 -87.714495        member   
3439725  41.857901 -87.668745  41.857506 -87.645991        casual   
4820877  41.893914 -87.641736  41.879255 -87.639904        member   
3249873  41.857618 -87.619411  41.884728 -87.619521        casual   

         ride_time[s]  
1711206       981.683  
1856538       233.435  
2730535       613.149  
369839        416.914  
5542980       208.615  
820333       4760.406  
5034074       134.170  
3439725       298.424  
4820877       525.659  
3249873      1429.328

Checking if there are rides with ride time less than 0. It may be because of error in sourcing the data or other application glitch

In [69]:
df_ride[df_ride['ride_time[s]'] < 0]

ride_id  rideable_type              started_at  \
3657369  A3C86101B94DCF83   classic_bike 2024-11-03 01:53:34.934   
3659357  9E3880C1B94EAB39  electric_bike 2024-11-03 01:48:07.988   
3662262  A21F9A727675300F  electric_bike 2024-11-03 01:53:22.726   
3669198  61C2AE038C0D67AB   classic_bike 2024-11-03 01:25:33.243   
3687068  F147973D700AA375  electric_bike 2024-11-03 01:50:02.506   
3703323  F600DA96D6308ABF  electric_bike 2024-11-03 01:55:22.919   
3705150  CFED9E2BF0A7F676  electric_bike 2024-11-03 01:50:18.246   
3713730  BB4F8E728A4714E3   classic_bike 2024-11-03 01:55:10.885   
3718110  5DD72265826A6FC0   classic_bike 2024-11-03 01:53:34.793   
3752418  CB6B83A69815601E   classic_bike 2024-11-03 01:25:14.083   
3755392  BCD7AB1DE9ACBAB0   classic_bike 2024-11-03 01:48:42.787   
3765633  B4DFEDC4D2BCE650   classic_bike 2024-11-03 01:24:08.760   
3765634  539D35F4E7FE1D13   classic_bike 2024-11-03 01:27:08.699   
3765635  CC11A50F0125A6B8   classic_bike 2024-11-03 01:26:02.313   
3778290  205F5CAF36F92EBF  electric_bike 2024-11-03 01:59:08.999   
3791212  143F5D35258E8E3F  electric_bike 2024-11-03 01:57:10.531   
3804651  5698BC27B89298D2  electric_bike 2024-11-03 01:48:08.185   
3818496  54E0B99065907777  electric_bike 2024-11-03 01:39:41.272   
3846980  2CD1341FCC0AECA8   classic_bike 2024-11-03 01:47:53.346   
3855676  C2D0EA867C94CB05  electric_bike 2024-11-03 01:57:10.166   
3863733  FE757D0F98D5DE20   classic_bike 2024-11-03 01:58:57.043   
3864363  583714F53028385C  electric_bike 2024-11-03 01:44:26.689   
3902127  CF11AD1AA5499FD6   classic_bike 2024-11-03 01:57:59.221   
3930187  6E973218589BD118   classic_bike 2024-11-03 01:55:58.187   
3962627  4BD1F0EE07E79421   classic_bike 2024-11-03 01:51:30.818   
3969529  6D8CC446C1251515  electric_bike 2024-11-03 01:52:30.180   

                       ended_at                 start_station_name  \
3657369 2024-11-03 01:04:55.434            Western Ave & Walton St   
3659357 2024-11-03 01:08:20.603           Kingsbury St & Kinzie St   
3662262 2024-11-03 01:06:40.133            Western Ave & Walton St   
3669198 2024-11-03 01:21:30.713  DuSable Lake Shore Dr & Monroe St   
3687068 2024-11-03 01:03:37.247               Larrabee St & Oak St   
3703323 2024-11-03 01:00:36.080         Aberdeen St & Jackson Blvd   
3705150 2024-11-03 01:03:39.105               Larrabee St & Oak St   
3713730 2024-11-03 01:02:13.083           Clark St & Bryn Mawr Ave   
3718110 2024-11-03 01:11:24.271           Wilton Ave & Belmont Ave   
3752418 2024-11-03 01:21:26.779  DuSable Lake Shore Dr & Monroe St   
3755392 2024-11-03 01:04:50.681       Milwaukee Ave & Wabansia Ave   
3765633 2024-11-03 01:21:28.841  DuSable Lake Shore Dr & Monroe St   
3765634 2024-11-03 01:21:32.265  DuSable Lake Shore Dr & Monroe St   
3765635 2024-11-03 01:21:39.906  DuSable Lake Shore Dr & Monroe St   
3778290 2024-11-03 01:13:02.165           Laflin St & Cullerton St   
3791212 2024-11-03 01:22:58.936         Kedvale Ave & Peterson Ave   
3804651 2024-11-03 01:08:24.249           Kingsbury St & Kinzie St   
3818496 2024-11-03 01:08:34.751     California Ave & Milwaukee Ave   
3846980 2024-11-03 01:04:05.631      Greenview Ave & Fullerton Ave   
3855676 2024-11-03 01:05:39.411                  Clark St & Elm St   
3863733 2024-11-03 01:09:01.069             Woodlawn Ave & 55th St   
3864363 2024-11-03 01:02:41.956         Racine Ave & Fullerton Ave   
3902127 2024-11-03 01:02:33.679              Clark St & Berwyn Ave   
3930187 2024-11-03 01:03:34.688       Milwaukee Ave & Wabansia Ave   
3962627 2024-11-03 01:16:39.875          Clark St & Wrightwood Ave   
3969529 2024-11-03 01:06:44.950         Ashland Ave & Blackhawk St   

        start_station_id                 end_station_name end_station_id  \
3657369     KA1504000103        Damen Ave & Charleston St          13288   
3659357     KA1503000043           Clark St & Schiller St   TA1309000024   
3662262     KA1504000103       Kedzie Ave & Milwaukee Ave     

In [70]:
df_ride[df_ride['ride_time[s]'] < 0].shape

(26, 14)

There are 26 records with negative ride time. I'll delete the rows with negative ride time.

In [71]:
df_ride = df_ride[df_ride['ride_time[s]'] > 0]

Check if there are records with same starting and ending stations

In [72]:
df_same_se = df_ride[df_ride['start_station_id'] == df_ride['end_station_id']]
df_same_se.info()

<class 'pandas.core.frame.DataFrame'>
Index: 198846 entries, 841 to 5627967
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             198846 non-null  object        
 1   rideable_type       198846 non-null  object        
 2   started_at          198846 non-null  datetime64[ns]
 3   ended_at            198846 non-null  datetime64[ns]
 4   start_station_name  198846 non-null  object        
 5   start_station_id    198846 non-null  object        
 6   end_station_name    198846 non-null  object        
 7   end_station_id      198846 non-null  object        
 8   start_lat           198846 non-null  float64       
 9   start_lng           198846 non-null  float64       
 10  end_lat             198846 non-null  float64       
 11  end_lng             198846 non-null  float64       
 12  member_casual       198846 non-null  object        
 13  ride_time[s]        198846 non-

Rides with ride time of less than 60secs and having the same starting and ending station, can be incorrect logging/testing/completely irrelevant

In [73]:
df_same_se[df_same_se['ride_time[s]']<60]

ride_id  rideable_type              started_at  \
4051     688E327F30462ADC  electric_bike 2024-06-03 18:35:48.295   
4052     0961F29D7B9BD9A0  electric_bike 2024-06-03 18:38:49.079   
4053     311716DCB6A70BA3  electric_bike 2024-06-03 18:37:07.658   
4055     BBF4BBE10D989C5B  electric_bike 2024-06-21 16:27:37.040   
4388     B34075C30222A59C  electric_bike 2024-06-02 15:14:23.669   
...                   ...            ...                     ...   
5611081  7DB3D5090AB5EC0D  electric_bike 2025-05-02 16:01:45.338   
5615706  9823366D00B6FCC1  electric_bike 2025-05-12 16:42:00.207   
5618250  3BE7136445D74551  electric_bike 2025-05-22 08:53:18.430   
5626028  22C9199C74B200E3  electric_bike 2025-05-27 23:33:56.926   
5627259  EB96B41DA5DA19C4  electric_bike 2025-05-11 10:18:09.220   

                       ended_at                         start_station_name  \
4051    2024-06-03 18:36:06.150               Washtenaw Ave & Peterson Ave   
4052    2024-06-03 18:39:16.506               Washtenaw Ave & Peterson Ave   
4053    2024-06-03 18:37:27.705               Washtenaw Ave & Peterson Ave   
4055    2024-06-21 16:27:53.573      Public Rack - Kedvale Ave & 63rd St E   
4388    2024-06-02 15:14:24.573               Washtenaw Ave & Peterson Ave   
...                         ...                                        ...   
5611081 2025-05-02 16:01:48.940             California Ave & Milwaukee Ave   
5615706 2025-05-12 16:42:01.913  Public Rack - Milwaukee Ave & Higgins Ave   
5618250 2025-05-22 08:53:58.809                Campbell Ave & Montrose Ave   
5626028 2025-05-27 23:34:34.700  Public Rack - Milwaukee Ave & Higgins Ave   
5627259 2025-05-11 10:18:44.487             California Ave & Milwaukee Ave   

        start_station_id                           end_station_name  \
4051                 458               Washtenaw Ave & Peterson Ave   
4052                 458               Washtenaw Ave & Peterson Ave   
4053                 458               Washtenaw Ave & Peterson Ave   
4055                 764      Public Rack - Kedvale Ave & 63rd St E   
4388                 458               Washtenaw Ave & Peterson Ave   
...                  ...                                        ...   
5611081            13084             California Ave & Milwaukee Ave   
5615706           1121.0  Public Rack - Milwaukee Ave & Higgins Ave   
5618250            15623                Campbell Ave & Montrose Ave   
5626028           1121.0  Public Rack - Milwaukee Ave & Higgins Ave   
5627259            13084             California Ave & Milwaukee Ave   

        end_station_id  start_lat  start_lng    end_lat    end_lng  \
4051               458  41.990702 -87.697064  41.990660 -87.697075   
4052               458  41.990613 -87.697010  41.990660 -87.697075   
4053               458  41.990663 -87.697113  41.990660 -87.697075   
4055               764  41.778574 -87.726207  41.778768 -87.725957   
4388               458  41.990657 -87.697035  41.990660 -87.697075   
...                ...        ...        ...        ...        ...   
5611081          13084  41.922695 -87.697153  41.922695 -87.697153   
5615706         1121.0  41.968596 -87.761209  41.968596 -87.761209   
5618250          15623  41.961524 -87.691177  41.961524 -87.691177   
5626028         1121.0  41.968653 -87.761018  41.968596 -87.761209   
5627259          13084  41.922695 -87.697153  41.922695 -87.697153   

        member_casual  ride_time[s]  
4051           casual        17.855  
4052           casual        27.427  
4053           casual        20.047  
4055           member        16.533  
4388           casual         0.904  
...               ...           ...  
5611081        member         3.602  
5615706        casual         1.706  
5618250        member        40.379  
5626028        casual        37.774  
5627259        casual        35.267  

[23037 rows x 14 columns]

23037 out of 198846 (11.6%) of the rides with same source and destination have ride time of less than 60secs

Removing such rides from the final data

In [82]:
df_clean = df_ride[~((df_ride['start_station_id'] == df_ride['end_station_id']) & (df_ride['ride_time[s]'] < 60))]
df_clean.shape

(3936597, 14)

In [83]:
df_final = df_clean.copy(deep=True)
df_final['Weekday'] = df_final['started_at'].dt.weekday

Renaming the columns for analysis

In [84]:
new_columns = {
    'rideable_type':'BikeType',
    'started_at': 'RideStart',
    'ended_at': 'RideEnd',
    'ride_time[s]': 'RideTime[s]',
    'start_station_name': 'StartStation',
    'end_station_name': 'EndStation',
    'member_casual': 'UserType'
}

df_final.rename(columns=new_columns, inplace=True)


In [85]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3936597 entries, 841 to 5628726
Data columns (total 15 columns):
 #   Column            Dtype         
---  ------            -----         
 0   ride_id           object        
 1   BikeType          object        
 2   RideStart         datetime64[ns]
 3   RideEnd           datetime64[ns]
 4   StartStation      object        
 5   start_station_id  object        
 6   EndStation        object        
 7   end_station_id    object        
 8   start_lat         float64       
 9   start_lng         float64       
 10  end_lat           float64       
 11  end_lng           float64       
 12  UserType          object        
 13  RideTime[s]       float64       
 14  Weekday           int32         
dtypes: datetime64[ns](2), float64(5), int32(1), object(7)
memory usage: 465.5+ MB
